src/snapshot/src/lib.rs

```diff
@@ -0,0 +1,921 @@
+        )?;
+        let format_version = validate_magic_id(magic_id).unwrap();
+        let hdr: SnapshotHdr =
+            SnapshotHdr::deserialize(&mut reader, &format_version_map, format_version)?;
+        let mut sections = HashMap::new();
+
+        for _ in 0..hdr.section_count {
+            let section = Section::deserialize(&mut reader, &format_version_map, format_version)?;
+            sections.insert(section.name.clone(), section);
+        }
+
+        Ok(Snapshot {
+            version_map,
+            hdr,
+            format_version,
+            sections,
+            // Not used when loading a snapshot.
+            target_version: 0,
+        })
+    }
+
+    /// Attempts to loads an existing snapshot and validate CRC.
+    pub fn load_with_crc64<T>(reader: &mut T, version_map: VersionMap) -> Result<Snapshot>
+    where
+        T: Read,
+    {
+        let mut crc_reader = CRC64Reader::new(reader);
+        let format_vm = Self::format_version_map();
+
+        // Read entire buffer in memory.
+        let snapshot = Snapshot::load(&mut crc_reader, version_map)?;
+        let computed_checksum = crc_reader.checksum();
+        let stored_checksum: u64 = Versionize::deserialize(&mut crc_reader, &format_vm, 0)?;
+
+        if computed_checksum != stored_checksum {
+            println!(
+                "Computed = {}, stored = {}",
+                computed_checksum, stored_checksum
+            );
+            return Err(Error::Crc64);
+        }
+
+        Ok(snapshot)
+    }
+
+    /// Saves a snapshot and include a CRC64 checksum.
+    pub fn save_with_crc64<T>(&mut self, writer: &mut T) -> Result<()>
+    where
+        T: std::io::Write,
+    {
+        let mut crc_writer = CRC64Writer::new(writer);
+        self.save(&mut crc_writer)?;
+
+        let checksum = crc_writer.checksum();
+        checksum.serialize(&mut crc_writer, &Self::format_version_map(), 0)?;
+        Ok(())
+    }
+
+    /// Save a snapshot.
+    pub fn save<T>(&mut self, mut writer: &mut T) -> Result<()>
+    where
+        T: std::io::Write,
+    {
+        self.hdr = SnapshotHdr {
+            data_version: self.target_version,
+            section_count: self.sections.len() as u16,
+        };
+
+        let format_version_map = Self::format_version_map();
+        let magic_id = build_magic_id(format_version_map.latest_version());
+
+        // Serialize magic id using the format version map.
+        magic_id.serialize(&mut writer, &format_version_map, 0 /* unused */)?;
+        // Serialize header using the format version map.
+        self.hdr.serialize(
+            &mut writer,
+            &format_version_map,
+            format_version_map.latest_version(),
+        )?;
+
+        // Serialize all the sections.
+        for section in self.sections.values() {
+            // The sections are already serialized.
+            section.serialize(
+                &mut writer,
+                &format_version_map,
```

Comment:
No. This is actually correct since we are searilizing a format specific structure `Section`. It already contains the state serialized at self.target_version in the `data` field (Vec<u8>)